element-view/src/androidMain/kotlin/ElementView.kt

```diff
@@ -0,0 +1,176 @@
+            state.value = state.value.copy(root = RootElement(), width = width, height = height)
+        }
+
+        /**
+         * Attach this to a view (which is itself attached to a window).
+         * This will immediately render the current state if possible.
+         */
+        internal fun onAttached(view: ElementView) {
+            state.value = State(view, RootElement(), view.width, view.height)
+            renderScope = object : CoroutineScope {
+                private val job = Job()
+                override val coroutineContext: CoroutineContext get() = job
+            }.also { launchRenderingIn(it) }
+        }
+
+        /** Detach this from a view (usually because the view was detached from a window). */
+        internal fun onDetached() {
+            renderScope = null
+            state.value = EMPTY_STATE
+        }
+
+        private fun launchRenderingIn(scope: CoroutineScope) = scope.launch {
+            // Unlimited buffer looks scary, but collection has way higher throughput than emission, so this doesn't grow large.
+            val buffer = MutableSharedFlow<Bitmap>(extraBufferCapacity = Int.MAX_VALUE)
+            val pool = BitmapPool()
+            scope.launch(Dispatchers.Main.immediate) {
+                buffer.collect { bitmap ->
+                    this@Adapter.bitmap?.let { pool.release(it) }
+                    this@Adapter.bitmap = bitmap
+                    state.value.view?.invalidate()
+                }
+            }
+            state.collectLatest { state ->
+                if (state.view == null) return@collectLatest
+                if (state.width == 0 || state.height == 0) return@collectLatest
+                val scalingFactor = applyDimension(COMPLEX_UNIT_DIP, 1f, state.view.resources.displayMetrics)
+                dataSource.conflate().collect { data ->
+                    updater.update(state.root, state.width / scalingFactor, state.height / scalingFactor, data)
+                    val bitmap = pool.acquire(state.width, state.height)
+                    state.root.draw(AndroidKanvas(state.view.context, Canvas(bitmap), scalingFactor))
+                    buffer.emit(bitmap)
+                }
+            }
+        }
+    }
+}
+
+/** Cache of no-longer-used [Bitmap] that might be useful in the future, to avoid allocating on every draw. */
+private class BitmapPool(
+    private val config: Bitmap.Config = Bitmap.Config.ARGB_8888,
+) {
+    /** Pixel count of the most recently requested bitmap. Used to clear old entries from the pool when an [ElementView] grows. */
+    private var size: Long = 0
+    private val mutex = Mutex()
+    private val pool = ArrayDeque<Bitmap>()
+
+    suspend fun acquire(width: Int, height: Int): Bitmap = mutex.withLock {
+        val newSize = width.toLong() * height
+        if (size > newSize) {
+            pool.forEach { it.recycle() }
+            pool.clear()
+        }
+        size = newSize
```

Comment:
Am I following this correctly?: If the requested pixel count is less than the previously used pixel count, then we clear the pool?

Why do we clear the pull if we requested a bitmap _smaller_ than the previously used bitmap?

Reply:
Good ol' fashioned catching bugs in PR, thanks!